### ISSUE TITLE
web2.dav.fileop methods should return deferreds also

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1839]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#1839|
|Type|enhancement|
|Created|2006-06-21 19:43:12Z|

Still have a few methods that need to return or handle deferreds, either those in the fileop module, those their callers.

<details><summary>Searchable metadata</summary>

```
trac-id__1839 1839
type__enhancement enhancement
reporter__wsanchez wsanchez
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__dav dav
time__1150918992000000 1150918992000000
changetime__1151088452000000 1151088452000000
version__None None
owner__ 
cc__wsanchez
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

branch:fileop-deferreds-1839

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @dreid|
|-|-|

OK, so now delete(), copy(), move(), put(), and mkcollection() return deferreds, the methods calling them handle this, and the tests still pass.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @wsanchez|
|-|-|

Handling of symlinks appears to be completely untested it shouldn't take very much effort to test them so please do that while you're in there mucking about.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

Lack of tests for symlink noted in bug [#1841](https://github.com/twisted/twisted/issues/1841).

The rest of the code passes on full-2.3, full-2.4, osx, and win32, so you can merge.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `closed`|
|-|-|

(In [17335]) Merged fileop-deferreds-1839
Reviewed by dreid
Fixes [#1839](https://github.com/twisted/twisted/issues/1839)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

Reverted, win32-select buildslave failed two tests:

```
===============================================================================
[ERROR]: twisted.web2.dav.test.test_copy.COPY.test_COPY_create

Traceback (most recent call last):
  File "c:\buildslave\win32-select\W32-full2.4-select\Twisted\twisted\web2\dav\test\util.py", line 92, in tearDown
    rmdir(self.docroot)
  File "c:\buildslave\win32-select\W32-full2.4-select\Twisted\twisted\web2\dav\fileop.py", line 491, in rmdir
    os.remove(os.path.join(dir, filename))
exceptions.OSError: [Errno 13] Permission denied: 'c:\\buildslave\\win32-select\\W32-full2.4-select\\Twisted\\_trial_temp\\twisted.web2.dav.test.test_copy\\COPY\\test_COPY_create\\jx5442\\temp\\test_mkcol.pyc'
===============================================================================
[ERROR]: twisted.web2.dav.test.test_copy.COPY.test_COPY_overwrite

Traceback (most recent call last):
  File "c:\buildslave\win32-select\W32-full2.4-select\Twisted\twisted\web2\dav\test\util.py", line 92, in tearDown
    rmdir(self.docroot)
  File "c:\buildslave\win32-select\W32-full2.4-select\Twisted\twisted\web2\dav\fileop.py", line 491, in rmdir
    os.remove(os.path.join(dir, filename))
exceptions.OSError: [Errno 13] Permission denied: 'c:\\buildslave\\win32-select\\W32-full2.4-select\\Twisted\\_trial_temp\\twisted.web2.dav.test.test_copy\\COPY\\test_COPY_overwrite\\fdma-0\\temp\\test_mkcol.pyc'
===============================================================================
```

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @exarkun<br>@wsanchez set status to `new`|
|-|-|

Fixed on the same branch by closing a FileStream's file ASAP during COPY.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @wsanchez|
|-|-|

 * twisted/web2/dav/test/test_copy.py - three unused imports: shutil, Deferred, and joinURL.  Not new in this branch, just pointing it out (particularly since another module seems to have had some import cleanup here)
 * twisted/web2/dav/method/copymove.py - os, deferredGenerator, waitForDeferred, StopTraversal, davxml, delete, ErrorResponse, and parentForURL imported but unused
 * twisted/web/dav/method/mkcol.py
   * os, Failure, davxml, statusForFailure imported but unused.
   * line 81 uses an unbound name, f
 * twisted/web2/dav/fileop.py
   * DeferredList imported but unused
   * line 243 uses an unbound name, errors
   * The delete function's docstring has a type - "approprite".  Same for the move, put, and mkcollection docstrings.
   * Including the waitForDeferred.getResult() call inside another expression looks like it might lead to problems to me.  The getResult really needs to be right after the yield; hiding it amongst other code seems like it might lead to mistakes later on.
   * The assert at the end of put() is dangerous.  It could be made to fail by some external interaction that this code has no control over.  Unit tests should cover the code well enough to be confident that put() is actually writing the correct file to the filesystem; beyond that, this code shouldn't try to assert this.  If it is really important for the file to be there at that point in time, some other exception than AssertionError should be raised if it is discovered to be missing.
 * twisted/web2/dav/resource.py - itertools, parser, IDAVPrincipalResource, joinURL, and parentForURL imported but unused.

Rest of the branch looks good.  Tests pass on my Ubuntu box and on the Win32 buildslave (I noticed your run and did another just to be safe).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

OK, these are all straightforward...

 * r17353 fixes a bunch of stuff caught by pyflakes.
 * r17354 fixes the "appropriate" typo
 * r17355 fixes the .getResult() concern; I think that in the expression, it's still called first, so we're OK, but I see how that's a little dodgy.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @exarkun|
|-|-|

Passed tests on my OS X box, and buildbot's full-2.3, full-2.4 and win32-select.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez commented|
|-|-|

r17361 removes the assert; you're right, someone else may have deleted the file.  I added that while double-checking the test cases, and it shouldn't have stayed in.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @wsanchez|
|-|-|

Changes look good!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `closed`|
|-|-|

(In [17362]) web2.dav.fileop methods should return deferreds also
Author: wsanchez
Reviewer: exarkun
Fixes [#1839](https://github.com/twisted/twisted/issues/1839)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
